### PR TITLE
Uart receiving data via dma and selecting speed and UART for logging output

### PR DESCRIPTION
**Pull request by Skorpionm**
# What's new

- furi_hal: add selecting speed and USART for logging output
- furi_hal: add USART1 receiving data via dma
- furi_hal: add LPUART1 receiving data via dma

# Verification 

- launch USB-UART Bridge. and check the reception of sending large uploads at high speeds

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/main/gpio/usb_uart_bridge.c [hunk 1]
```diff
@@ -75,13 +76,18 @@ static const CdcCallbacks cdc_cb = {
...
+    size_t data_len,
+    void* context) {
     UsbUartBridge* usb_uart = (UsbUartBridge*)context;
 
-    if(ev == UartIrqEventRXNE) {
-        furi_stream_buffer_send(usb_uart->rx_stream, &data, 1, 0);
-        furi_thread_flags_set(furi_thread_get_id(usb_uart->thread), WorkerEvtRxDone);
-    }
+    UNUSED(ev);
+    uint8_t data[data_len];
```
Comment (skotopes) [inline applications/main/gpio/usb_uart_bridge.c]: don't allocate data on stack with variable passed from outside

File: applications/main/gpio/usb_uart_bridge.c [hunk 2]
```diff
@@ -75,13 +76,18 @@ static const CdcCallbacks cdc_cb = {
...
+    furi_hal_uart_dma_rx(id_uart, data, data_len);
+    furi_stream_buffer_send(usb_uart->rx_stream, data, data_len, 100);
```
Comment (skotopes) [inline applications/main/gpio/usb_uart_bridge.c]: no timeouts in interrupts

File: firmware/targets/f7/furi_hal/furi_hal_uart.c [hunk 3]
```diff
@@ -198,47 +468,113 @@ void furi_hal_uart_tx(FuriHalUartId ch, uint8_t* buffer, size_t buffer_size) {
...
 }
 
-void USART1_IRQHandler(void) {
-    if(LL_USART_IsActiveFlag_RXNE_RXFNE(USART1)) {
-        uint8_t data = LL_USART_ReceiveData8(USART1);
-        irq_cb[FuriHalUartIdUSART1](UartIrqEventRXNE, data, irq_ctx[FuriHalUartIdUSART1]);
-    } else if(LL_USART_IsActiveFlag_ORE(USART1)) {
-        LL_USART_ClearFlag_ORE(USART1);
+size_t furi_hal_uart_dma_rx(FuriHalUartId ch, uint8_t* data, size_t len) {
+    furi_assert(uart[ch] != NULL && uart[ch]->buffer_rx_ptr != NULL);
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_uart.c:505]: FURI_IS_IRQ_MODE check

File: firmware/targets/f7/furi_hal/furi_hal_uart.c [hunk 4]
```diff
@@ -198,47 +468,113 @@ void furi_hal_uart_tx(FuriHalUartId ch, uint8_t* buffer, size_t buffer_size) {
...
-        if(ch == FuriHalUartIdUSART1) {
-            NVIC_DisableIRQ(USART1_IRQn);
-            LL_USART_DisableIT_RXNE_RXFNE(USART1);
-        } else if(ch == FuriHalUartIdLPUART1) {
-            NVIC_DisableIRQ(LPUART1_IRQn);
-            LL_LPUART_DisableIT_RXNE_RXFNE(LPUART1);
-        }
-        irq_cb[ch] = cb;
-        irq_ctx[ch] = ctx;
+size_t furi_hal_uart_dma_bytes_available(FuriHalUartId ch) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_uart.c]: private

**Review by skotopes — CHANGES_REQUESTED** [review]

File: applications/main/gpio/usb_uart_bridge.c (hunk 1 above)
Comment (Skorpionm) [reply applications/main/gpio/usb_uart_bridge.c]: fix

File: applications/main/gpio/usb_uart_bridge.c (hunk 2 above)
Comment (Skorpionm) [reply applications/main/gpio/usb_uart_bridge.c]: fix

File: firmware/targets/f7/furi_hal/furi_hal_uart.c (hunk 3 above)
Comment (Skorpionm) [reply firmware/targets/f7/furi_hal/furi_hal_uart.c:505]: fix

File: firmware/targets/f7/furi_hal/furi_hal_uart.c (hunk 4 above)
Comment (Skorpionm) [reply firmware/targets/f7/furi_hal/furi_hal_uart.c]: fix

**Pull request closed without merging**